upstairs/src/volume.rs

```diff
@@ -245,44 +252,147 @@ impl Volume {
+    // If a volume has a read only parent, we do the work to read from
+    // the read only side, and write_unwritten to the LBA of the SubVolume
+    // that is "below" it.
+    //
+    // Still TODO: here
+    // If there is an error, how to we tell Nexus?
+    // If we finish the scrub, how do we tell Nexus?
     pub fn scrub(&self) -> Result<(), CrucibleError> {
         println!("Scrub check for {}", self.uuid);
         // XXX Can we assert volume is activated?
 
+        // Do I need ROP to be ARC?
         if let Some(ref read_only_parent) = self.read_only_parent {
-            let ts = read_only_parent.total_size()?;
-            let bs = read_only_parent.get_block_size()? as usize;
-
-            println!("Scrub for total_size:{:?} block_size:{:?}", ts, bs);
+            let scrub_start = Instant::now();
+            println!("Scrub for {} begins", self.uuid);
+
+            let bs = read_only_parent.get_block_size().unwrap() as usize;
+            println!(
+                "Scrub with total_size:{:?} block_size:{:?}",
+                read_only_parent.total_size().unwrap(),
+                bs
+            );
             let start = read_only_parent.lba_range.start;
             let end = read_only_parent.lba_range.end;
-            println!("Scrub will go from {:?} to {:?}", start, end);
 
-            for offset in start..end {
+            // Based on some seat of the pants measurements, we are doing
+            // 256 KiB IOs during the scrub. Select how many blocks this is.
+            // TODO: Determine if this value should be adjusted.
+            let mut block_count = 262144 / bs;
+            println!(
+                "Scrub from block {:?} to {:?} in ({}) {:?} size IOs",
+                start,
+                end,
+                block_count,
+                block_count * bs,
+            );
+
+            let mut errors = 0;
+            let mut retries = 0;
+            let showstep = (end - start) / 25;
+            let mut showat = start + showstep;
+            let mut offset = start;
+            while offset < end {
+                if offset + block_count as u64 > end {
+                    block_count = (end - offset) as usize;
+                    println!(
+                        "Adjust block_count to {} at offset {}",
+                        block_count, offset
+                    );
+                }
+                assert!(offset + block_count as u64 <= end);
                 let block = Block::new(offset, bs.trailing_zeros());
-                let buffer = Buffer::new(bs);
+                let buffer = Buffer::new(block_count * bs);
+
+                // The read will first try to establish a connection to
+                // the remote side before returning something we can block_wait
+                // on.  If that initial connection fails, we can retry.
+                let mut retry_needed = true;
+                let mut retry_count = 0;
+                while retry_needed {
```

Comment:
Yeah, I agree with your worry :)
The reason I have one here is that I was frequently seeing this connection fail when testing on folgers.  It was not
the actual block_wait that was failing, but that initial connection to catacomb that seemed to have trouble making
a connection.  Once I got the connection (and could wait on the waiter) things worked and I never had an issue.

Maybe it's something with the way the read to url path conflicts with itself (as the instance is usually booting while the scrub is going), but without this retry I could almost never get through a larger scrub.

If there is a "real" problem (and yes, that's in quotes), the block_wait will fail, and for that there is no retry, so it should flow back to Nexus, when that plumbing is written.

Reply:
> The reason I have one here is that I was frequently seeing this connection fail when testing on folgers. It was not the actual block_wait that was failing, but that initial connection to catacomb that seemed to have trouble making a connection. Once I got the connection (and could wait on the waiter) things worked and I never had an issue.

I'd love to dig into this! At first glance, I don't understand why this is happening when the instance boots ok. What errors did you see?

Reply:
I saw it when I had the scrubber doing block size IOs, which means it was doing a lot more connections to the URL read path.  I don't have the error any longer, but when I get folgers back I can reduce the IO size and try again to see what the message is.   IIRC, it was always in the initial connection part of the `read` request,

this part I believe:
```
        let response = self
            .client
            .get(&self.url)
            .header(
                RANGE,
                format!(
                    "bytes={}-{}",
                    start,
                    start + data_vec.len() as u64 - 1
                ),  
            )       
            .send()
            .map_err(|e| CrucibleError::GenericError(e.to_string()))?;
```

So I never even got to wait on the header.

Reply:
The first thought that comes to mind: I wonder if this would be improved by changing the Upstairs to be `async`. Each read will be waiting on that `send()` to return and I don't know how reqwest is doing the blocking there. We might have to stick a `[tokio](https://docs.rs/tokio/0.2.4/tokio/index.html)::[task](https://docs.rs/tokio/0.2.4/tokio/task/index.html)::[block_in_place](https://docs.rs/tokio/0.2.4/tokio/task/fn.block_in_place.html)` in there?

Reply:
Here is the error I see:
scrub Error: error sending request for url (http://[fd00:1122:3344:101::1]:54321/media/cloud/debian-11-genericcloud-amd64.raw): operation timed out, offset 239514

The retry triggered and the next attempt worked.  I see this only when I changed the size of each IO to be 512 (block_size).

Reply:
I filed issue #452 for this